src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -98,64 +98,61 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
     // Function to expose currently registered controllers to test code only
     @testOnlyMethod()
     public registeredNotebookControllers(): VSCodeNotebookController[] {
-        return this.registeredControllers;
+        return Array.from(this.registeredControllers.values());
     }
 
     // Find all the notebook controllers that we have registered
     public async loadNotebookControllers(): Promise<void> {
         if (!this.controllersPromise) {
-            this.controllersPromise = this.loadNotebookControllersImpl()
-                .then((controllers) => {
-                    // Just assign here as this is our initial set of controllers
-                    // anything that adds or updates should make sure the initial load has happened first
-                    this.registeredControllers = controllers;
-                })
+            const stopWatch = new StopWatch();
+
+            this.controllersPromise = Promise.all([
+                this.loadNotebookControllersImpl(true),
+                this.loadNotebookControllersImpl(false)
```

Comment:
Fetch both in parallel, if non-python kernels finish first, then those kernels will be registerd early (validatd by adding a simple sleep of 10s in pyton discovery)